lib/nameresolver.go

```
@@ -0,0 +1,139 @@
+		logging.Log.Debug.Printf("Failed modify deployment %q: % #v", depPair.ID(), depPair.Post)
+		return nil, &DiffResolution{
+			DeploymentID: depPair.ID(),
+			Desc:         "not updated",
+			Error:        WrapResolveError(errors.Errorf("Unable to modify new deployment %q: no artifact for SourceID %q", depPair.ID(), depPair.Post.SourceID)),
+		}
+	}
+	return d, nil
+}
+
+// Stop always returns no error because we don't need a deploy artifact to delete a deploy
+func (names *nameResolver) Stop(dp *DeployablePair) (*DeployablePair, *DiffResolution) {
+	da := maybeResolveSingle(names.registry, dp.Prior)
+	return &DeployablePair{ExecutorData: dp.ExecutorData, name: dp.name, Prior: da, Post: nil}, nil
+}
+
+// Stable always returns no error because we don't need a deploy artifact for unchanged deploys
+func (names *nameResolver) Stable(dp *DeployablePair) (*DeployablePair, *DiffResolution) {
+	da := maybeResolveSingle(names.registry, dp.Post)
+	return &DeployablePair{ExecutorData: dp.ExecutorData, name: dp.name, Prior: da, Post: da}, nil
+}
+
+// XXX now that everything is DeployablePairs, this can probably be simplified
+
+func maybeResolveSingle(r Registry, dep *Deployable) *Deployable {
+	logging.Log.Vomit.Printf("Attempting to resolve optional artifact: %#v (stable or deletes don't need images)", dep)
+	da, err := resolveName(r, dep)
+	if err != nil {
+		logging.Log.Debug.Printf("Error resolving stopped or stable deployment (proceeding anyway): %#v: %#v", dep, err)
+	}
+	return da
+}
+
+func resolveName(r Registry, d *Deployable) (*Deployable, *DiffResolution) {
+	art, err := guardImage(r, d.Deployment)
+	if err != nil {
+		return d, &DiffResolution{
+			DeploymentID: d.ID(),
+			Error:        &ErrorWrapper{error: err},
+		}
+	}
+	d.BuildArtifact = art
+	return d, nil
+}
+
+func resolvePair(r Registry, depPair *DeployablePair) (*DeployablePair, *DiffResolution) {
+	prior, _ := resolveName(r, depPair.Prior)
```

I always get nervous when throwing errors away but I can see that the error here is operationally irrelevant. However, if an error is returned, I would rather at least modify `prior` to indicate that fetching it resulted in an error, to make any logs printing it out more explicit, e.g. `fmt.Sprintf("failed to fetch image name because %s", err)` I don't think it's worth logging by itself though. If modifying `prior` would have a negative impact, then at least a comment explaining why we are ignoring the error would be better than nothing.